### DISCUSSION TITLE
Fix/Improve Nullability Analysis in Loops

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#4955] [Language Ideas]
In the example below, ```ret``` should be definitely assigned since the loop is guaranteed to run.
```
#nullable enable

using System;

namespace ConsoleApp23 {
    class Program {
        static void Main(string[] args) {
            TestObject();
        }

        public static object TestObject() {
            var ret = default(object?);

            for (int i = 0; i < 1; i++) {
                ret = new();
            }
            return ret;
        }

    }
}
```

**Comment by huoyaoyuan** [comment]
> since the loop is guaranteed to run.

Analyzing arbitrary loop is impossible. The current behavior only handles constant `true` and `false`, which is fine.

If we have something like "range based loops", analysis for constant ranges is good. Treating `for` loop as range based loop is quite complex, since small modification can make the loop no longer recognized.

**Comment by HaloFour** [comment]
The definite assignment flow analysis in the compiler doesn't handle this kind of case either.  If you don't give `ret` a default value on declaration it will be a compiler error to attempt to return it from the method.
